### PR TITLE
[TSPS-5] Add jobs endpoints - post, get one, get all

### PR DESCRIPTION
**Pull request by mmorgantaylor**
This is a too-large PR to get in a bunch of work done to create a few new job endpoints, described in TSPS-5. (Future PRs will be smaller!)

Here I add:
- POST jobs/{pipeline_id} to request a job of the specified pipeline_id
- GET jobs/{pipeline_id} to retrieve a list of all jobs the user has requested of the specified pipeline_id
- GET jobs/{pipeline_id} to retrieve a specified job 

Note that testing coverage is incomplete and will be added in future tickets along with more error handling.

### CODE REVIEW

File: service/src/main/java/bio/terra/pipelines/app/controller/JobsApiController.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    String pipelineVersion = body.getPipelineVersion();
+
+    logger.info(
+        "Creating {} pipeline job (version {}) for {} subject {}",
+        pipelineId,
+        pipelineVersion,
+        userRequest.getEmail(),
+        userId);
+
+    // TODO revisit the following (this is copy-pasta from WSM)
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/app/controller/JobsApiController.java:71]: update this TODO - we WILL need to check user permissions

File: service/src/main/java/bio/terra/pipelines/service/JobsService.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  private final JobsDao jobsDao;
+
+  @Autowired
+  public JobsService(JobsDao jobsDao) {
+    this.jobsDao = jobsDao;
+  }
+
+  public UUID createJob(String userId, JobRequest jobRequest) {
+    UUID jobId = createJobId();
+    Timestamp timeSubmitted = getCurrentTimestamp();
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/JobsService.java:29]: add TODO: in future get the timestamp from CBAS/Cromwell, don't generate it here

File: service/src/main/java/bio/terra/pipelines/service/JobsService.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    logger.info("Create new {} job with job_id {}", jobRequest.getPipelineId(), jobId);
+
+    // placeholder for actually doing something; for now we're just writing the info to the database
+    //        JobBuilder createJob = jobService ...
+
+    String status = "SUBMITTED";
+    Job jobFull =
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/JobsService.java:37]: add a comment to note that this class will grow (like DbJob)

File: service/src/main/java/bio/terra/pipelines/service/JobsService.java (hunk 2 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/service/JobsService.java:29]: actually, we want to keep this - this is the time we received the request and started the full process (of which the CBAS submission is only a part)

File: service/src/main/java/bio/terra/pipelines/service/JobsService.java (hunk 3 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/service/JobsService.java:37]: can use something like this to reference the db object https://stackoverflow.com/questions/17496038/javadoc-link-to-method-in-other-class

File: service/src/main/java/bio/terra/pipelines/db/JobsDao.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            .addValue("status", job.getStatus(), VARCHAR);
+    try {
+      jdbcTemplate.update(sql, params);
+      logger.info("Inserted record for job {}", jobUuid);
+    } catch (DuplicateKeyException e) {
+      String message = e.getMessage();
+      if (message != null
+          && message.contains("duplicate key value violates unique constraint \"jobs_pkey\"")) {
+        // Job with job_id already exists.
+        throw new DuplicateObjectException(
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/db/JobsDao.java:77]: add TODO: in this case, don't throw an error, rather retry with new UUID. make a new ticket for this!

File: common/openapi.yml [hunk 5]
```diff
@@ -136,44 +222,123 @@ components:
...
+        jobId:
+          $ref: "#/components/schemas/JobId"
+
+    JobId:
+      description: |
+        Unique identifier (UUID) for a job request.
+      type: string
+      format: string
+
+    JobResponse:
```
Comment (mmorgantaylor) [inline common/openapi.yml:246]: rename JobGetResponse

File: common/openapi.yml [hunk 6]
```diff
@@ -136,44 +222,123 @@ components:
...
+
+    CreateJobRequestBody:
+      description: |
+        Object containing the user-provided information about a job request.
+      type: object
+      properties:
+        pipelineVersion:
+          $ref: "#/components/schemas/PipelineVersion"
+
+    CreatedJob:
```
Comment (mmorgantaylor) [inline common/openapi.yml:234]: rename to `PostJobResponse`

File: common/openapi.yml [hunk 7]
```diff
@@ -136,44 +222,123 @@ components:
                   type: string
 
     # Please keep alphabetized
-    TspsPipeline:
+
+    CreateJobRequestBody:
```
Comment (mmorgantaylor) [inline common/openapi.yml:226]: rename to `PostJobRequestBody`

File: service/src/main/java/bio/terra/pipelines/app/controller/JobsApiController.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    return apiResult;
+  }
+
+  static ApiJobsGetResponse jobsToApi(List<Job> jobList) {
+    ApiJobsGetResponse apiResult = new ApiJobsGetResponse();
+
+    for (Job job : jobList) {
+      var apiJob = jobToApi(job);
+
+      apiResult.add(apiJob); // is there a better function to use here? e.g. addJobItem()
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/app/controller/JobsApiController.java:126]: TODO: look at this - find the example from template

File: service/src/main/java/bio/terra/pipelines/db/JobsDao.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+   *
+   * @param userId unique identifier of the calling user
+   * @param pipelineId identifier of the pipeline
+   * @param jobId UUID identifier of the job
+   * @return Job object
+   */
+  public Job getJob(String userId, String pipelineId, String jobId) {
+    DbJob dbJob =
+        getDbJobIfExists(userId, pipelineId, jobId)
+            .orElseThrow(() -> new JobNotFoundException(String.format("Job {} not found.", jobId)));
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/db/JobsDao.java:100]: functionally check that this is working

File: service/src/main/java/bio/terra/pipelines/app/controller/GlobalExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+  // -- Error Report - one of our exceptions --
+  @ExceptionHandler(ErrorReportException.class)
+  public ResponseEntity<ApiErrorReport> errorReportHandler(ErrorReportException ex) {
+    return buildApiErrorReport(ex, ex.getStatusCode(), ex.getCauses());
+  }
+
+  // -- validation exceptions - we don't control the exception raised
+  @ExceptionHandler({
+    MethodArgumentNotValidException.class,
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/app/controller/GlobalExceptionHandler.java:36]: add JobNotFoundException method here

Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/app/controller/GlobalExceptionHandler.java:36]: (make a follow up ticket for this)

File: service/src/main/java/bio/terra/pipelines/app/controller/GlobalExceptionHandler.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            + ". Ensure that all types are correct and that enums have valid values.";
+    ApiErrorReport errorReport =
+        new ApiErrorReport()
+            .message(validationErrorMessage)
+            .statusCode(HttpStatus.BAD_REQUEST.value());
+    return new ResponseEntity<>(errorReport, HttpStatus.BAD_REQUEST);
+  }
+
+  // Exception thrown by Spring Retry code when interrupted.
+  @ExceptionHandler({BackOffInterruptedException.class})
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/app/controller/GlobalExceptionHandler.java:60]: note we're not actively using Spring Retry yet - we don't use @Retryable yet - can note this for now, we will likely use it later

File: service/src/main/java/bio/terra/pipelines/iam/AuthenticatedUserRequest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
+package bio.terra.pipelines.iam;
+
+import bio.terra.common.exception.ApiException;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import java.util.Optional;
+import java.util.UUID;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthenticatedUserRequest {
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/iam/AuthenticatedUserRequest.java:10]: this isn't being used now - can we delete these two classes for now? (this and AuthenticatedUserRequestFactory.java)

File: service/src/main/java/bio/terra/pipelines/service/JobsService.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+    return jobsDao.createJob(jobFull);
+  }
+
+  private UUID createJobId() {
+    return UUID.randomUUID();
+  }
+
+  private Timestamp getCurrentTimestamp() {
+    // TODO add time zone
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/JobsService.java:55]: add ticket reference here

File: service/src/main/java/bio/terra/pipelines/service/model/Job.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        .setUserId(dbJob.userId())
+        .setPipelineId(dbJob.pipelineId())
+        .setPipelineVersion(dbJob.pipelineVersion())
+        .setTimeSubmitted(dbJob.timeSubmitted())
+        .setTimeCompleted(dbJob.timeCompleted())
+        .setStatus(dbJob.status())
+        .build();
+  }
+
+  public static class Builder {
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/model/Job.java:74]: consider removing this Builder entirely

File: service/src/main/java/bio/terra/pipelines/service/model/JobRequest.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+  public String getPipelineId() {
+    return pipelineId;
+  }
+
+  public String getPipelineVersion() {
+    return pipelineVersion;
+  }
+
+  public static class Builder {
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/model/JobRequest.java:24]: remove Builder

File: service/src/main/java/bio/terra/pipelines/service/model/JobRequest.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
+package bio.terra.pipelines.service.model;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
+@JsonDeserialize(builder = JobRequest.Builder.class)
+public class JobRequest {
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/model/JobRequest.java:6]: remove this class and just use Job

File: service/src/main/java/bio/terra/pipelines/service/model/JobRequest.java (hunk 15 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/service/model/JobRequest.java:24]: deleting the entire class instead

File: service/src/main/java/bio/terra/pipelines/app/controller/JobsApiController.java (hunk 8 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/app/controller/JobsApiController.java:123]: I couldn't actually find this in either the template project or TPS. so I'll leave it as is (and remove the comment)

File: service/src/main/java/bio/terra/pipelines/db/JobsDao.java (hunk 9 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/db/JobsDao.java:101]: it works! 
`{
  "message": "Job {} not found.",
  "statusCode": 404,
  "causes": []
}
`

Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/db/JobsDao.java:101]: fixed
`{
  "message": "Job foo not found.",
  "statusCode": 404,
  "causes": []
}`

**Review by koncheto-broad — APPROVED** [review]
LGTM!  We can take on the created tickets piecemeal to target the areas commented on in this PR.

**Pull request merged**